### PR TITLE
tetrio-desktop: 8.0.0 -> 9.0.0

### PR DESCRIPTION
**Pull request by Kropatz**
## Description of changes
Updated the download link and references to the executable file.
Patch notes: https://tetr.io/about/patchnotes/#chlog_6_4_3

## Things done
Played a couple of games, works perfectly.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Kropatz** [comment]
@wackbyte for review? Sorry, don't know how this works

**Review by wackbyte — APPROVED** [review]
LGTM, I'll try building it later

**Comment by wackbyte** [comment]
Result of `nixpkgs-review pr 296454` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>tetrio-desktop</li>
  </ul>
</details>

File: pkgs/games/tetrio-desktop/default.nix
Comment (drupol) [inline pkgs/games/tetrio-desktop/default.nix:70]:
```suggestion
      --replace-fail "Exec=/opt/TETR.IO/TETR.IO" "Exec=\"$out/opt/TETR.IO/TETR.IO\""
```

**Review by drupol — CHANGES_REQUESTED** [review]
Hi !

Thanks for your contribution.

I've left some comments, let me know if this is clear enough.

Don't forget to add `meta.mainProgram` too.

**Comment by Kropatz** [comment]
Thanks for the comment. I added `meta.mainProgram` and learned from reading the docs this affects nix run. So I tested it by running `NIXPKGS_ALLOW_UNFREE=1 nix run --impure .#tetrio-desktop`.
It launched the app successfully

**Review by drupol — COMMENTED** [review]
It LGTM now.

What could be improved in the future:

1. Don't use `sha256`, use `hash` instead
2. I wonder how we should keep backward compatibility. Since the executable name has changed, it is currently broken. Don't you think we should add an entry in aliases to let people know about that name change?

**Review by wackbyte — APPROVED** [review]

**Comment by drupol** [comment]
@wackbyte What's your point of view regarding backward compatibility ?

**Comment by wackbyte** [comment]
I agree that this is a breaking change and that something should be done. (However, I don't have experience with nixpkgs' procedure for this)

**Comment by drupol** [comment]
I just added an alias, can you please review and let me know if it's OK ?

File: pkgs/top-level/aliases.nix
Comment (Kropatz) [inline pkgs/top-level/aliases.nix:1147]: This won't work because there is no "TETR.IO" package, it is still called `tetrio-desktop`.
Also I think the package naming convention is to have it all lowercase, so `tetr.io`. 

Presonally I don't really see a reason why we should rename the package though :thinking:

Comment (wackbyte) [reply pkgs/top-level/aliases.nix:1147]: I don't see why the package should be renamed either. If it is renamed, it seems odd to me for the stylization of the name to be preserved in the package name. `tetrio` would be preferable to me.

Comment (wackbyte) [reply pkgs/top-level/aliases.nix:1147]: My concern for backwards compatibility was about the name of the binary (rather than that of the package) changing from `tetrio-desktop` to `TETR.IO`.

Comment (drupol) [reply pkgs/top-level/aliases.nix]: Right, I undo my commit.

**Review by wackbyte — APPROVED** [review]

**Comment by drupol** [comment]
Let's rename the executable to `tetrio`

**Comment by drupol** [comment]
Here's my view of this thing:

```nix
{ stdenv
, lib
, fetchurl
, dpkg
, autoPatchelfHook
, wrapGAppsHook
, alsa-lib
, cups
, libGL
, libX11
, libXScrnSaver
, libXtst
, mesa
, nss
, gtk3
, libpulseaudio
, systemd
}:

let
  libPath = lib.makeLibraryPath [
    libGL
    libpulseaudio
    systemd
  ];
in
stdenv.mkDerivation (finalAttrs: {
  pname = "tetrio-desktop";
  version = "9.0.0";

  src = fetchurl {
    url = "https://tetr.io/about/desktop/builds/${lib.versions.major finalAttrs.version}/TETR.IO%20Setup.deb";
    hash = "sha256-UriLwMB8D+/T32H4rPbkJAy/F/FFhNpd++0AR1lwEfs=";
  };

  nativeBuildInputs = [
    dpkg
    autoPatchelfHook
    wrapGAppsHook
  ];

  dontWrapGApps = true;

  buildInputs = [
    alsa-lib
    cups
    libX11
    libXScrnSaver
    libXtst
    mesa
    nss
    gtk3
  ];

  unpackCmd = "dpkg -x $curSrc src";

  installPhase = ''
    runHook preInstall

    mkdir -p $out/bin
    cp -r opt/ usr/share/ $out
    ln -s $out/opt/TETR.IO/TETR.IO $out/bin/tetrio

    substituteInPlace $out/share/applications/TETR.IO.desktop \
      --replace-fail "Exec=/opt/TETR.IO/TETR.IO" "Exec=tetrio"

    runHook postInstall
  '';

  postFixup = ''
    wrapProgram $out/opt/TETR.IO/TETR.IO \
      --prefix LD_LIBRARY_PATH : ${libPath}:$out/opt/TETR.IO \
      ''${gappsWrapperArgs[@]}
  '';

  meta = {
    downloadPage = "https://tetr.io/about/desktop/";
    description = "TETR.IO desktop client";
    homepage = "https://tetr.io";
    license = lib.licenses.unfree;
    longDescription = ''
      TETR.IO is a modern yet familiar online stacker.
      Play against friends and foes all over the world, or claim a spot on the leaderboards - the stacker future is yours!
    '';
    mainProgram = "tetrio";
    maintainers = with lib.maintainers; [ wackbyte ];
    platforms = [ "x86_64-linux" ];
  };
})
```

Tetrio plus has been removed since it's not supported on version 9.

I also tested to download the `.tar.gz`  instead, but there's no desktop file in it sadly.

@wackbyte let me know if you're OK with this version. I tested it locally without any trouble.

**Comment by drupol** [comment]
I pushed a commit including all of this, let me know if you like it.

File: pkgs/by-name/te/tetrio-desktop/package.nix
Comment (Kropatz) [inline pkgs/by-name/te/tetrio-desktop/package.nix:65]: Desktop file didn't work for me with this. (because tetrio is not on the path)
```suggestion
      --replace-fail "Exec=/opt/TETR.IO/TETR.IO" "Exec=\"$out/bin/tetrio\""
```

Comment (drupol) [reply pkgs/by-name/te/tetrio-desktop/package.nix]: Weird, since `tetrio` is in `$out/bin`, it should have been added to the path, no ?
Anyway, I updated the PR. Let me know if it's better.

Comment (drupol) [reply pkgs/by-name/te/tetrio-desktop/package.nix]: (The Desktop File specification doesn't add quotes around properties: https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#example)

**Comment by Kropatz** [comment]
By the way, I tested the newest CI artifact from the tetrio plus gitlab and that worked. 
But there is no official release yet, so I don't know how we should handle that.
I guess we just wait until it releases and make a PR then?

File: pkgs/by-name/te/tetrio-desktop/package.nix
Comment (Kropatz) [reply pkgs/by-name/te/tetrio-desktop/package.nix]: Good to know, thanks :+1:

**Comment by drupol** [comment]
> By the way, I tested the newest CI artifact from the tetrio plus gitlab [here](https://gitlab.com/UniQMG/tetrio-plus/-/jobs/6465395934/artifacts/raw/app.asar.zip) and that worked. But there is no official release yet, so I don't know how we should handle that. I guess we just wait until it releases and make a PR then?

I guess that's the way to go.

**Comment by Kropatz** [comment]
Oh another thing, when we remove the `withTetrioPlus` flag, does that break the existing configuration of people that have it set to true? Or does nix just do nothing without erroring? 
I don't know how to test that.

**Comment by drupol** [comment]
> Oh another thing, when we remove the `withTetrioPlus` flag, does that break the existing configuration of people that have it set to true? Or does nix just do nothing without erroring? I don't know how to test that.

True! Nice catch.

Let's see with @wackbyte (the maintainer) and have its point of view.

**Comment by Kropatz** [comment]
I manged to test it like this:
```sh
export NIX_PATH=current-pr=/projects/github/nixpkgs:$NIX_PATH  
NIXPKGS_ALLOW_UNFREE=1 nix-build -E 'with (import <current-pr>{}); tetrio-desktop.override { withTetrioPlus = true; }'
```

And that throws an error about the unexpected argument.
```
error:
       … while calling a functor (an attribute set with a '__functor' attribute)

         at «string»:1:31:

            1| with (import <current-pr>{}); tetrio-desktop.override { withTetrioPlus = true; }
             |                               ^

       … while calling a functor (an attribute set with a '__functor' attribute)

         at /projects/github/nixpkgs/lib/customisation.nix:104:43:

          103|       # Re-call the function but with different arguments
          104|       overrideArgs = mirrorArgs (newArgs: makeOverridable f (overrideWith newArgs));
             |                                           ^
          105|       # Change the result of the function call by applying g to it

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: function 'anonymous lambda' called with unexpected argument 'withTetrioPlus'

       at /projects/github/nixpkgs/pkgs/by-name/te/tetrio-desktop/package.nix:1:1:

            1| { stdenv
             | ^
            2| , lib
```

I'd say we keep the flag but just make it do nothing, and add a comment explaining why.

**Comment by Kropatz** [comment]
Added the dummy flag, LGTM now

**Comment by drupol** [comment]
> Added the dummy flag, LGTM now

In that case I would also add the other parameter that goes with it.

**Comment by Kropatz** [comment]
I don't know how to do that

**Comment by drupol** [comment]
```
tetrio-plus ? throw "tetrio-plus is not supported on version 9"
```

**Comment by Kropatz** [comment]
Doesn't that just set the default value though? It wouldn't do anything because it's unused.

File: pkgs/by-name/te/tetrio-desktop/package.nix
Comment (drupol) [inline pkgs/by-name/te/tetrio-desktop/package.nix:19]: I actually think this is not a so good idea.

Comment (wackbyte) [reply pkgs/by-name/te/tetrio-desktop/package.nix:19]: What do you think is wrong?

Comment (drupol) [reply pkgs/by-name/te/tetrio-desktop/package.nix:19]: Because if it's not set, the throw will never be evaluated. And if it's set, it won't be used later on in the file. Therefore, the `throw` is totally useless there.

Comment (wackbyte) [reply pkgs/by-name/te/tetrio-desktop/package.nix:19]: Oh, I didn't see the `postInstall` that used it was removed. Should that be added back?

Comment (drupol) [reply pkgs/by-name/te/tetrio-desktop/package.nix:19]: No, I don't think so.

Comment (wackbyte) [reply pkgs/by-name/te/tetrio-desktop/package.nix:19]: So those params should be removed for now, and everything can be re-added once TETR.IO Plus is working?

Comment (drupol) [reply pkgs/by-name/te/tetrio-desktop/package.nix:19]: Yes, that's what I would do.

Comment (wackbyte) [reply pkgs/by-name/te/tetrio-desktop/package.nix:19]: Sounds good to me.

Comment (drupol) [reply pkgs/by-name/te/tetrio-desktop/package.nix:19]: Feel free to adjust this PR, I won't have time to do it until tonight.

Comment (Kropatz) [reply pkgs/by-name/te/tetrio-desktop/package.nix:19]: I don't get it why though? It would break the configs of everyone who has `withTetrioPlus` set. Why not leave it as a dummy parameter that does nothing for the time being? When tetrio plus is working again, they all get it back automatically.

Comment (wackbyte) [reply pkgs/by-name/te/tetrio-desktop/package.nix:19]: That's also true. If the `postInstall` is re-added, users who have it enabled will see an error that it's not currently supported, but could still provide their own nightly build of it. Silently removing the option doesn't sound so good. At the very least, if it stays removed for now, a warning should be emitted when `withTetrioPlus` is enabled.

Comment (drupol) [reply pkgs/by-name/te/tetrio-desktop/package.nix]: Ready to merge for you then?

**Review by wackbyte — APPROVED** [review]
I'd ultimately prefer the binary to not be renamed (to stay as `TETR.IO` rather than `tetrio`) for consistency's sake, but I don't mind it this way. LGTM

**Review by drupol — APPROVED** [review]

**Pull request merged**
